src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+                break;
+
+            const auto & match = matches[sort_node];
+
+            //std::cerr << "====== Finding match for " << sort_column_node->result_name << ' ' << static_cast<const void *>(sort_column_node) << std::endl;
+
+            if (match.node && match.node == sort_column_node)
+            {
+                //std::cerr << "====== Found direct match" << std::endl;
+
+                /// We try to find the match first even if column is fixed. In this case, potentially more keys will match.
+                /// Example: 'table (x Int32, y Int32) ORDER BY x + 1, y + 1'
+                ///          'SELECT x, y FROM table WHERE x = 42 ORDER BY x + 1, y + 1'
+                /// Here, 'x + 1' would be a fixed point. But it is reasonable to read-in-order.
+
+                current_direction = descr.direction;
+                if (match.monotonicity)
+                {
+                    current_direction *= match.monotonicity->direction;
+                    strict_monotonic = match.monotonicity->strict;
+                }
+
+                ++next_descr_column;
+                ++next_sort_key;
+            }
+            else if (fixed_key_columns.contains(sort_column_node))
+            {
+                //std::cerr << "+++++++++ Found fixed key by match" << std::endl;
+                ++next_sort_key;
+            }
+            else
+            {
+
+                //std::cerr << "====== Check for fixed const : " << bool(sort_node->column) << " fixed : " << fixed_columns.contains(sort_node) << std::endl;
+                bool is_fixed_column = sort_node->column || fixed_columns.contains(sort_node);
+                if (!is_fixed_column)
+                    break;
+
+                order_key_prefix_descr.push_back(descr);
+                ++next_descr_column;
+            }
+        }
+
+        /// read_direction == 0 means we can choose any global direction.
+        /// current_direction == 0 means current key if fixed and any direction is possible for it.
+        if (current_direction && read_direction && current_direction != read_direction)
+            break;
+
+        if (read_direction == 0)
+            read_direction = current_direction;
+
+        if (current_direction)
+            order_key_prefix_descr.push_back(descr);
+
+        if (current_direction && !strict_monotonic)
+            break;
+    }
+
+    if (read_direction == 0 || order_key_prefix_descr.empty())
+        return nullptr;
+
+    return std::make_shared<InputOrderInfo>(order_key_prefix_descr, next_sort_key, read_direction, limit);
+}
+
+InputOrderInfoPtr buildInputOrderInfo(
+    ReadFromMergeTree * reading,
+    const FixedColumns & fixed_columns,
+    const ActionsDAGPtr & dag,
+    const SortDescription & description,
+    size_t limit)
+{
+    const auto & sorting_key = reading->getStorageMetadata()->getSortingKey();
+    const auto & sorting_key_columns = sorting_key.column_names;
+
+    return buildInputOrderInfo(
```

Comment:
Probably if we decide to split this function call in multiple lines, better to have each function argument on the same line.

Reply:
Or on the different lines.